### ISSUE TITLE
Lazy broadcast flattening

### ISSUE THREAD

**Issue by ChrisRackauckas** [rdeits/Flatten.jl#2]
The idea is simple. The implementation is complex.

**Comment by rdeits** [comment]
BTW, all the fun new work is happening over at this fork: https://github.com/rafaqz/Flatten.jl 

@rafaqz since your fork is way ahead of this old project, perhaps you'd like to turn on GitHub issues for the fork? Or I'm happy to merge in your changes here.

**Comment by rafaqz** [comment]
@rdeits wasn't sure where to go with that, I'm still wondering wether people generally even like or need the MetaFields tricks. I can put in a pull request, it just depends if you want part in maintaining it in future. I'll turn on comments for now.

@ChrisRackauckas I refactored the original code quite a bit, but also added a lot more complexity for optional field handling and metafield flattening with MetaFields.jl, and also Units.jl handling. I'm not sure it could be simplified much further while remaining type stable and fast.

**Comment by rafaqz** [comment]
@ChrisRackauckas for structs where all fields are Float64 etc and you want them all in the vecto my fork adds a lot of useless cruft, and there might be simpler options. (Although in terms of speed it will all just compile out). 

I imagined it to be more of a tool for people using DiffEq tools with complex structures in their models and fields that shouldn't be in the vector, which isn't possible in this original form of Flatten.jl or https://github.com/simonbyrne/PackVec.jl.

**Comment by ChrisRackauckas** [comment]
>@ChrisRackauckas for structs where all fields are Float64 etc and you want them all in the vecto my fork adds a lot of useless cruft, and there might be simpler options. (Although in terms of speed it will all just compile out).

That's just a `FieldVector`. 

http://juliaarrays.github.io/StaticArrays.jl/latest/pages/api.html#FieldVector-1

I'm looking for heterogeneous to handle nestings like PackVec. PackVec.jl was actually built for this purpose, but we are going to replace PackVec.jl since I want something that can more generally lazy broadcast without having to flatten into a vector. IMO flattening should just happen via `convert(Vector,...)` via the indexing structure that allows broadcast to work.

**Comment by rafaqz** [comment]
I think maybe were talking past each other here. My version of flatten lets you reconstruct a struct from a vector and fields on an existing struct that _arent in the vector_. And vis a vis, build a vector excluding some of the fields at any point in the struct heriarchy. And update that set of fields to be whichever you want. Basic flattening is relatively simple and I agree it should just be convert(). 

Excluding fields needs an additional mechanism, but I doubt you need it your for purposes.

**Comment by ChrisRackauckas** [comment]
Yeah, I don't need field exclusion. Sounds like we are looking for something related but different and should probably build to our own use cases.

**Comment by rafaqz** [comment]
Ok this is far more API-like and generalised:

https://github.com/rafaqz/Flatten.jl/blob/master/src/Flatten.jl#L34

`field_expressions` handles all flattening/reconstruction processes,
Everything after those 25 lines is just padding out the actual functionality.

Field exclusion is now a passed-in function name, so you can use this to ignore it:
`flatten_all(x, field) = Flat()`

The conditionals are still there in the generated function, but they compile out. This could mean that both requirements are met
with a small amount of code. Although I think  theres a bit more editing and thought needed.
